shared/utility/strategy/soccer_strategy.hpp

```diff
@@ -0,0 +1,124 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef UTILITY_STRATEGY_SOCCER_STRATEGY_HPP
+#define UTILITY_STRATEGY_SOCCER_STRATEGY_HPP
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+
+#include "message/input/RoboCup.hpp"
+#include "message/localisation/Ball.hpp"
+#include "message/localisation/Field.hpp"
+#include "message/strategy/TeamMates.hpp"
+
+
+namespace utility::strategy {
+
+    using message::localisation::Ball;
+    using message::localisation::Field;
+    using message::strategy::TeamMates;
+
+    struct Possession {
+        enum Value { SELF, TEAMMATE, OPPONENT, NONE };
+        Value value = Value::SELF;
+    };
+
+    struct Robots {
+        // Robot's distance to ball
+        double distance_to_ball = 0.0;
+        // Id of the robot
+        uint id = 0;
+        // Instance of Possession struct, indicating if the robot has the ball, is close, etc.
+        Possession possession;
+    };
+
+    std::vector<Robots> get_sorted_bots(const Ball& ball, const TeamMates& teammates, const Field& field) {
+        std::vector<Robots> robots;
+
+        // Transforms the ball's position from world coordinates to field coordinates using field.Hfw.
+        // rBWw is the ball's position in world coordinates.
+        Eigen::Vector3d rBFf = field.Hfw * ball.rBWw;
+
+        // Looping through teammmates, for each teammate, a 'Robots' object is created.
+        // Robot's position (mate.rRFf) is retrieved, and distance from robot to ball is calculated using norm.
+        // Then is stored in 'distance_to_ball'
+        // Each 'Robots' object is added to the robots vector.
+        for (const auto& mate : teammates.teammates) {
+            Robots robot;
+            robot.id               = mate.id;
+            Eigen::Vector3d rRFf   = mate.rRFf;
+            robot.distance_to_ball = (rRFf - rBFf).norm();
+            robots.push_back(robot);
+        }
+
+        // Robot's are sorted by distance to the ball, if there is a tie, they are sorted by robot id.
+        std::sort(robots.begin(), robots.end(), [](const Robots& a, const Robots& b) {
+            return (a.distance_to_ball < b.distance_to_ball)
+                   || (a.distance_to_ball == b.distance_to_ball && a.id < b.id);
+        });
+
+        // Sorted list of robots is returned.
+        return robots;
+    }
+
+
+    Possession get_possession(const Ball& ball, const TeamMates& teammates, const Field& field, double threshold) {
+
+        // Function determines who has possession based on proximity and a threshold distance.
+        // First calls 'get_sorted_bots()' to get the list of robots sotered by distance to ball.
+        // If no robot's are found, it returns 'Possession::NONE'
+        std::vector<Robots> sortedRobots = get_sorted_bots(ball, teammates, field);
+        if (sortedRobots.empty()) {
+            return Possession{Possession::NONE};
+        }
```

Comment:
this is a really good start to using the sorted robots, but then the rest of the code below still seems to be the old stuff and doesn't use the sorted robots?